espnet2/layers/augmentation.py

```diff
@@ -0,0 +1,535 @@
+        window = window_func(win_length, dtype=waveform.dtype, device=waveform.device)
+    ret = torchaudio.functional.pitch_shift(
+        waveform,
+        sample_rate,
+        n_steps,
+        bins_per_octave=bins_per_octave,
+        n_fft=n_fft,
+        win_length=win_length,
+        hop_length=hop_length,
+        window=window,
+    )
+    return ret
+
+
+def speed_perturb(waveform, sample_rate: int, factor: float):
+    """Speed perturbation which also changes the pitch.
+
+    Note: This function should be used with caution as it changes the signal duration.
+
+    Args:
+        waveform (torch.Tensor): audio signal (..., time)
+        sample_rate (int): sampling rate in Hz
+        factor (float): speed factor (e.g., 0.9 for 90% speed)
+        lengths (torch.Tensor): lengths of the input signals
+
+    Returns:
+        ret (torch.Tensor): perturbed signal (..., time)
+    """
+    orig_freq = sample_rate
+    source_sample_rate = int(factor * orig_freq)
+    target_sample_rate = int(orig_freq)
+
+    gcd = math.gcd(source_sample_rate, target_sample_rate)
+    source_sample_rate = source_sample_rate // gcd
+    target_sample_rate = target_sample_rate // gcd
+
+    ret = torchaudio.functional.resample(
+        waveform, source_sample_rate, target_sample_rate
+    )
+    return ret
+
+
+def time_stretch(
+    waveform,
+    sample_rate: int,
+    factor: float,
+    n_fft: float = 0.032,
+    win_length: Optional[float] = None,
+    hop_length: float = 0.008,
+    window: Optional[str] = "hann",
+):
+    """Time scaling (speed up in time without modifying pitch) via phase vocoder.
+
+    Note: This function should be used with caution as it changes the signal duration.
+
+    Args:
+        waveform (torch.Tensor): audio signal (..., time)
+        sample_rate (int): sampling rate in Hz
+        factor (float): speed-up factor (e.g., 0.9 for 90% speed and 1.3 for 130% speed)
+        n_fft (int): length of FFT (in second)
+        win_length (int or None): The window length (in second) used for STFT
+            If None, it is treated as equal to n_fft
+        hop_length (int): The hop size (in second) used for STFT
+        window (str or None): The windowing function applied to the signal after
+            padding with zeros
+
+    Returns:
+        ret (torch.Tensor): perturbed signal (..., time)
+    """
+    n_fft = int(sample_rate * n_fft)
+    if hop_length is None:
+        hop_length = n_fft // 4
+    else:
+        hop_length = int(sample_rate * hop_length)
+    if win_length is None:
+        win_length = n_fft
+    if window is not None:
+        window_func = getattr(torch, f"{window}_window")
+        window = window_func(win_length, dtype=waveform.dtype, device=waveform.device)
+    spec = torch.stft(
+        waveform, n_fft, hop_length, win_length, window=window, return_complex=True
+    )
+    freq = spec.size(-2)
+    phase_advance = torch.linspace(0, math.pi * hop_length, freq)[..., None]
```

Comment:
is [..., None] equivalent to .unsqueeze(-1) here?

Reply:
Yes. These are just the same operations.